programs/psitest/main.cpp

```diff
@@ -757,22 +933,58 @@ struct callbacks
       return 0;
    }
 
-   uint32_t testerWriteFile(int32_t fd, span<const char> content)
+   uint32_t testerWasi_fd_write(wasi_fd_t fd, 
+                                span<const wasi_iovec_t> iovs, 
+                                wasm_ptr<wasi_size_t> nwritten)
    {
-      auto* file = get_file(fd);
-      if (!file)
-         return wasi_errno_badf;
-      if (!content.size() || fwrite(content.data(), content.size(), 1, file->f) == 1)
-         return 0;
-      return wasi_errno_io;
+      *nwritten = 0;
+
+      for (auto iovs_item : iovs)
+      {
+         auto* file = get_file(fd);
+         if (!file)
+            return wasi_errno_badf;
```

Comment:
Done, and I've done the same with the fread function below. Can you please check?

Reply:
It shouldn't fail just because iovs is empty. The reason I mentioned iovs being empty is that moving the check outside the loop causes the check to be executed when iovs is empty, when it would not before (Something that always needs to be considered for loop invariant code motion). I was trying to point out that this change in behavior is correct.